src/typescript/frontend/src/lib/aptos-indexer/fungible-assets.ts

```diff
@@ -0,0 +1,115 @@
+
+export type TokenBalance = {
+  amount: string;
+  asset_type: string;
+  owner_address: string;
+  metadata: {
+    symbol: string;
+    decimals: number;
+  };
+};
+
+export type FetchEmojicoinBalancesResponse = {
+  current_fungible_asset_balances: TokenBalance[];
+};
+
+export type FetchFungibleAssetsParams = {
+  ownerAddress?: string;
+  // Can be used to fetch holders of specific coin.
+  assetType?: string;
+  offset?: number;
+  limit?: number;
+};
+
+export async function fetchFungibleAssetsBalance({
+  ownerAddress,
+  assetType,
+  offset = 0,
+  limit = 100,
+}: FetchFungibleAssetsParams): Promise<FetchEmojicoinBalancesResponse> {
+  const conditions = ['{ metadata: {token_standard: {_eq: "v1"}} }', '{ amount: {_gt: "0"} }'];
+
+  if (assetType) conditions.push("{ asset_type: {_eq: $assetType} }");
+  else conditions.push('{ asset_type: {_ilike: "%coin_factory::Emojicoin%"} }');
+
+  if (ownerAddress) {
+    conditions.push("{ owner_address: {_eq: $ownerAddress} }");
+  }
+
+  return aptosClient.queryIndexer({
+    query: {
+      query: `
+        query CoinsData($ownerAddress: String, $assetType: String, $offset: Int, $limit: Int) {
+          current_fungible_asset_balances(
+            where: {
+              _and: [${conditions.join(",")}]
+            }
+            offset: $offset
+            limit: $limit
+          ) {
+            owner_address
+            amount
+            asset_type
+            metadata {
+              decimals
+              symbol
+            }
+          }
+        }`,
+      variables: {
+        ownerAddress,
+        offset,
+        limit,
+        assetType,
+      },
+    },
+  });
+}
+
+export async function fetchAllFungibleAssetsBalance({
+  ownerAddress,
+  assetType,
+  // Fetch at most 1000 by default, which corresponds to 10 requests.
+  max = 1000,
+}: { max?: number } & Omit<FetchFungibleAssetsParams, "offset" | "limit">): Promise<
+  TokenBalance[]
+> {
+  const limit = 100;
+  let offset = 0;
+  let allTokens: TokenBalance[] = [];
+  let hasMore = true;
+
+  while (hasMore && offset < max) {
+    const response = await fetchFungibleAssetsBalance({ ownerAddress, offset, limit, assetType });
+    const tokens = response.current_fungible_asset_balances;
+    allTokens = [...allTokens, ...tokens];
+
+    // If we got less than the limit, we've reached the end.
+    hasMore = tokens.length === limit;
+    if (hasMore) {
+      offset += limit;
+    }
+  }
+
+  // If specific assetType provided, no need to filter out.
+  if (assetType) {
+    return allTokens;
+  }
+
+  // Filter out non-emojicoin tokens.
+  return allTokens.filter((token) => {
+    // Return early if symbol is not existing emoji.
+    if (!(token.metadata.symbol in SYMBOL_EMOJIS)) return;
+    const address = getEmojicoinMarketAddressAndTypeTags({
+      symbolBytes: encodeEmojis([...token.metadata.symbol]),
+    });
+    return address.emojicoin.toString() === token.asset_type;
```

Comment:
```suggestion
    return address.emojicoin.toString() === parseTypeTag(token.asset_type).toString();
```

I think you need to standardize the `token.asset_type` here (unless I missed it elsewhere above)

It's basically non-deterministic whether or not you're getting a standardized (no leading zeros) address back from the graphQL endpoint because it's difficult to know whether which "address-like" values are normalized by the TS SDK.

Basically, assume it's *not* standardized, otherwise you'll get a false negative here for things like `0x000000abc...::coin_factory::Emojicoin`